File: ch003_stack/leetcode_20.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println(isValid("()[]{}"))
}

func isValid(s string) bool {
	stack := make([]rune, 0)
    pairs := map[rune]rune{
        ')': '(',
        ']': '[',
        '}': '{',
    }
	for _, v := range s {
        if strings.Contains("([{", string(v)) {
            stack = append(stack, v)
            continue
        }

        if len(stack) == 0 {
            return false
        }   
        
        if val, ok := pairs[v]; ok {
            if stack[len(stack)-1] != val {
                return false
            }
            stack = stack[:len(stack)-1]
        }
	}
	return len(stack) == 0
}

File: ch003_stack/leetcode_20.py
def isValid(s: str) -> bool:
        stack = []
        pairs = {')': '(', ']': '[', '}': '{'}
        for value in s:
            if value in pairs.values():
                stack.append(value)
                continue
            if len(stack) == 0:
                return False
            if pairs.get(value, None) != stack.pop():
                return False
        return len(stack) == 0

print(isValid("[]{}()[{}]"))